File: packages/gavio-py/gavio/providers/mock.py
"""MockProvider — deterministic, offline provider for dev mode and tests."""

from __future__ import annotations

import hashlib
import time
from collections.abc import AsyncIterator

from ..pricing import PricingProvider, estimate_tokens
from ..request import GavioRequest
from ..response import GavioResponse
from ..types import TokenUsage
from .base import ProviderAdapter


class MockProvider(ProviderAdapter):
    """Returns a canned response without any network call.

    If ``response`` is None, it echoes the last user message so the pipeline
    (including PII restore) is observable end to end.
    """

    def __init__(
        self,
        response: str | None = None,
        model_version: str = "mock-1",
        pricing: PricingProvider | None = None,
    ) -> None:
        super().__init__(pricing)
        self._response = response
        self._model_version = model_version

    @property
    def provider_name(self) -> str:
        return "mock"

    def _content_for(self, request: GavioRequest) -> str:
        if self._response is not None:
            return self._response
        last_user = next(
            (m.get("content", "") for m in reversed(request.messages) if m.get("role") == "user"),
            "",
        )
        return f"[mock reply] {last_user}"

    async def complete(self, request: GavioRequest) -> GavioResponse:
        started = time.monotonic()
        content = self._content_for(request)
        usage = TokenUsage(
            prompt_tokens=estimate_tokens(request.prompt_text()),
            completion_tokens=estimate_tokens(content),
        )
        return self._build_response(request, content, usage, self._model_version, started)

    async def embed(self, request: GavioRequest) -> GavioResponse:
        """Deterministic 8-dim vector per message content (F-SEC-10)."""
        started = time.monotonic()
        vectors = [_mock_vector(m.get("content", "")) for m in request.messages]
        usage = TokenUsage(prompt_tokens=estimate_tokens(request.prompt_text()))
        return self._build_embed_response(request, vectors, usage, self._model_version, started)

    async def stream(self, request: GavioRequest) -> AsyncIterator[str]:
        for token in self._content_for(request).split(" "):
            yield token + " "

    async def health_check(self) -> bool:
        return True

    @property
    def reported_model_version(self) -> str | None:
        return self._model_version


def _mock_vector(text: str, dims: int = 8) -> list[float]:
    """Stable pseudo-embedding: sha256 bytes scaled to [0, 1)."""
    digest = hashlib.sha256(text.encode("utf-8")).digest()
    return [digest[i] / 255.0 for i in range(dims)]

File: packages/gavio-js/src/providers/mock.ts
/** mockProvider — deterministic, offline provider for dev mode and tests. */

import { createHash } from 'node:crypto'

import { PricingProvider, estimateTokens } from '../pricing.js'
import type { GavioRequest } from '../request.js'
import type { GavioResponse } from '../response.js'
import { TokenUsage } from '../types.js'
import { BaseProviderAdapter } from './base.js'

export interface MockProviderOptions {
  response?: string | null
  modelVersion?: string
  pricing?: PricingProvider
}

/**
 * Returns a canned response without any network call.
 *
 * If `response` is null/undefined, it echoes the last user message so the
 * pipeline (including PII restore) is observable end to end.
 */
class MockProvider extends BaseProviderAdapter {
  private readonly response: string | null
  private readonly modelVersion: string

  constructor(options: MockProviderOptions = {}) {
    super(options.pricing)
    this.response = options.response ?? null
    this.modelVersion = options.modelVersion ?? 'mock-1'
  }

  get providerName(): string {
    return 'mock'
  }

  override get reportedModelVersion(): string | null {
    return this.modelVersion
  }

  private contentFor(request: GavioRequest): string {
    if (this.response !== null) return this.response
    const lastUser = [...request.messages]
      .reverse()
      .find((m) => m.role === 'user')
    return `[mock reply] ${lastUser?.content ?? ''}`
  }

  async complete(request: GavioRequest): Promise<GavioResponse> {
    const started = performance.now()
    const content = this.contentFor(request)
    const usage = new TokenUsage(
      estimateTokens(request.promptText()),
      estimateTokens(content),
    )
    return this.buildResponse(request, content, usage, this.modelVersion, started)
  }

  /** Deterministic 8-dim vector per message content (F-SEC-10). */
  async embed(request: GavioRequest): Promise<GavioResponse> {
    const started = performance.now()
    const vectors = request.messages.map((m) => mockVector(m.content ?? ''))
    const usage = new TokenUsage(estimateTokens(request.promptText()))
    return this.buildEmbedResponse(request, vectors, usage, this.modelVersion, started)
  }

  async *stream(request: GavioRequest): AsyncIterable<string> {
    for (const token of this.contentFor(request).split(' ')) {
      yield token + ' '
    }
  }

  async healthCheck(): Promise<boolean> {
    return true
  }
}

/** Stable pseudo-embedding: sha256 bytes scaled to [0, 1). */
function mockVector(text: string, dims = 8): number[] {
  const digest = createHash('sha256').update(text, 'utf-8').digest()
  return Array.from({ length: dims }, (_, i) => digest[i]! / 255)
}

/** Factory: build a mock provider adapter. */
export function mockProvider(options: MockProviderOptions = {}): ProviderAdapterMock {
  return new MockProvider(options)
}

export type ProviderAdapterMock = MockProvider
